{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 1: controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re  #Importing the module for regular expressions\n",
    "germ = open(\"Germplasm.tsv\", \"r\") #Opening the Germplasm file in read mode and assigning it to an object\n",
    "germ.seek(0)  #Taking the cursor to the beginning of the file\n",
    "germlist = []\n",
    "for line in germ.readlines()[1:]:    #iterate over each line in the file which are string, except over the first line, index 0, that are the headers\n",
    "    print(line)\n",
    "    matchobj = re.search( r'^(AT[0-9]G[0-9]+)', line)#do a regular expression search for the locus codes and save the results in matchobj\n",
    "    germlist.append(matchobj.group(1)) #I create a list with all the locus codes to be able to compare it with the list from locusgene\n",
    "    print(\"The AGI locus codes are\", matchobj.group(1)) #print the results of the search for each line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re \n",
    "loc = open(\"LocusGene.tsv\", \"r\")\n",
    "loclist = []\n",
    "for line in loc.readlines()[1:]:\n",
    "    matchobj = re.search( r'^(AT[0-9]G[0-9]+)', line)\n",
    "    loclist.append(matchobj.group(1))\n",
    "    print(\"The AGI locus codes are\", matchobj.group(1))\n",
    "loc.seek(0)   # set it back to the beginning again for this lesson..\n",
    "loclist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if germlist == loclist:\n",
    "    print(\"Both documents follow the same locus sequence\")\n",
    "else:\n",
    "    print(\"The documents do not follow the same sequence\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I performed a regular expression search over the databases to check if both locus codes where written equally and then append them to a list to be able to check if they followed they same order. \n",
    "In the last box of code we can see how the two lists created are equal, therefore the lists follow the same order."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2: Design and create the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql create database exam_week_2;\n",
    "%sql show databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use exam_week_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the previous lines of code I have loaded de sqlmagic extension and created the database that I will be using for this exam."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                                user='root',\n",
    "                                password='root',\n",
    "                                db='exam_week_2',  #the database is exam_week_2 as shown above\n",
    "                                charset='utf8mb4', \n",
    "                                cursorclass=pymysql.cursors.DictCursor)\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv') as csvfile:    \n",
    "            cursor.execute(\"drop table germplasm\")  #this line is just in case the base was created previously\n",
    "            germ = open(\"Germplasm.tsv\", \"r\") #open the tsv in read mode and save it in germ\n",
    "            germplasm = germ.readlines()      #reading the lines in the tsv via the germ object and saving them into the object germplasm\n",
    "            import re\n",
    "            mo = re.search( r'([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)[\\n]', germplasm[0]) #performing a regular expression search to obtain the columns of the header, the text we are reading is the first row of the document\n",
    "            colnames = [mo.group(1), mo.group(2), mo.group(3), mo.group(4)]   #saving the names of the columns in a list using the differents group of group method for mo\n",
    "            sql = \"CREATE TABLE germplasm( \" +colnames[0] + \" VARCHAR(12) NOT NULL PRIMARY KEY, \" +colnames[1] + \" VARCHAR(30) NOT NULL, \" + colnames[2] + \" VARCHAR(600) NOT NULL, \" + colnames[3] + \" INTEGER NOT NULL)\" #creating the sql tables with the header names from the colnames list\n",
    "            cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close() # if I close before I commit, the inserts are lostprint(colnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select * from germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import pymysql\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                                user='root',\n",
    "                                password='root',\n",
    "                                db='exam_week_2',\n",
    "                                charset='utf8mb4', # note utf8... this is important for unusual characters!\n",
    "                                cursorclass=pymysql.cursors.DictCursor)\n",
    "#onnection.autocommit = False\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            cursor.execute(\"drop table locus\")\n",
    "            loc = open(\"LocusGene.tsv\", \"r\")\n",
    "            locus = loc.readlines()\n",
    "            import re\n",
    "            mo = re.search( r'([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)[\\n]', locus[0])\n",
    "            colnames1 = [mo.group(1), mo.group(2), mo.group(3)]\n",
    "            sql = \"CREATE TABLE locus( \" +colnames1[0] + \" VARCHAR(20) NOT NULL PRIMARY KEY, \" +colnames1[1] + \" VARCHAR(30) NOT NULL, \" + colnames1[2] + \" INTEGER NOT NULL)\"\n",
    "            cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close() # if I close before I commit, the inserts are lostprint(colnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select * from locus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the lines above I have created the two tables from their respective tsv files using pymysql and regular expression searches"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv') as csvfile:   \n",
    "            res = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"') #I am saving the lines from Dictreader as rows in which the values are quoted and separated by tab\n",
    "            for row in res:   #I am iterating over the object created before which is a dictreader that maps the operation of each row into a dict\n",
    "                sql = \"\"\"INSERT INTO germplasm (Locus, germplasm, phenotype, pubmed) values       \n",
    "                ('{}', '{}','{}',{})\"\"\".format(row[colnames[0]], row[colnames[1]], row[colnames[2]],row[colnames[3]]) #I am iterating over the res object, obtaining rows which are dictionaries. Then extracting the values that are paired with the colnames keys\n",
    "                cursor.execute(sql)\n",
    "            connection.commit() \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', 17369351),\n",
       " ('AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401),\n",
       " ('AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', 17486125),\n",
       " ('AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744),\n",
       " ('AT2G03720', 'SALK_042433', 'Multiple straight hairs', 16367956),\n",
       " ('AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', 15215505),\n",
       " ('AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900),\n",
       " ('AT2G05210', 'pot1-1', 'No visible phenotype.', 17627276),\n",
       " ('AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837),\n",
       " ('AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180),\n",
       " ('AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444),\n",
       " ('AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653),\n",
       " ('AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', 10821278),\n",
       " ('AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843),\n",
       " ('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770),\n",
       " ('AT3G02870', 'vtc4-1', 'ascorbate deficient', 16595667),\n",
       " ('AT3G03260', 'hdg8-1', 'No visible phenotype.', 16778018),\n",
       " ('AT4G14790', 'pdd17', 'Defective pollen development.', 19237690),\n",
       " ('AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066),\n",
       " ('AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425),\n",
       " ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186),\n",
       " ('AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662),\n",
       " ('AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499),\n",
       " ('AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', 12747832),\n",
       " ('AT4G16480', 'atint4-2', 'No visible phenotype.', 16603666),\n",
       " ('AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', 18799749),\n",
       " ('AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', 17960244),\n",
       " ('AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216),\n",
       " ('AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401),\n",
       " ('AT5G11510', 'myb3r4-1', 'No visible phenotype.', 17287251),\n",
       " ('AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687),\n",
       " ('AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678)]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select * from germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            res = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            for row in res:\n",
    "                sql = \"\"\"INSERT INTO locus (Locus, Gene, ProteinLength) values       \n",
    "                ('{}', '{}',{})\"\"\".format(row[colnames1[0]], row[colnames1[1]], row[colnames1[2]]) #El de aquí es el que viene del csv y el de arriba el de las tablas sql\n",
    "                cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332),\n",
       " ('AT1G01060', 'LHY', 290),\n",
       " ('AT1G01140', 'CIPK9', 223),\n",
       " ('AT1G01220', 'FKGP', 190),\n",
       " ('AT2G03720', 'MRH6', 189),\n",
       " ('AT2G03800', 'GEK1', 196),\n",
       " ('AT2G04240', 'XERICO', 256),\n",
       " ('AT2G05210', 'POT1A', 221),\n",
       " ('AT3G02130', 'RPK2', 284),\n",
       " ('AT3G02140', 'TMAC2', 300),\n",
       " ('AT3G02230', 'RGP1', 301),\n",
       " ('AT3G02260', 'BIG', 279),\n",
       " ('AT3G02310', 'SEP2', 175),\n",
       " ('AT3G02680', 'NBS1', 190),\n",
       " ('AT3G02850', 'SKOR', 234),\n",
       " ('AT3G02870', 'VTC4', 311),\n",
       " ('AT3G03260', 'HDG8', 194),\n",
       " ('AT4G14790', 'SUV3', 312),\n",
       " ('AT4G15210', 'BAM5', 313),\n",
       " ('AT4G15560', 'DXS', 219),\n",
       " ('AT4G15570', 'MAA3', 294),\n",
       " ('AT4G15802', 'HSBP', 254),\n",
       " ('AT4G15880', 'ESD4', 265),\n",
       " ('AT4G16420', 'ADA2B', 279),\n",
       " ('AT4G16480', 'INT4', 284),\n",
       " ('AT5G10480', 'PAS2', 301),\n",
       " ('AT5G10510', 'AIL6', 310),\n",
       " ('AT5G11110', 'SPS2', 232),\n",
       " ('AT5G11260', 'HY5', 221),\n",
       " ('AT5G11510', 'MYB3R-4', 336),\n",
       " ('AT5G12200', 'PYD2', 310),\n",
       " ('AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select * from locus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have filled the tables using pymsql and basic for loops and iterable commands and indexation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "import io\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus;\"\"\" #I am creating the joined table\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()            #saving the table in the object results which is a list of dictionaries\n",
    "        report = \"\\t\".join(list(results[0].keys())) #creating the header for the document joining the keys of the first dict in the list, although I could have used any dict. And saving it in the object report\n",
    "        for dicc in results:      #iterating over the list, taking the dictionaris in the object dicc\n",
    "            linea = []            #creating a new list\n",
    "            for value in dicc.values():   #iterating over the values in each dictionary\n",
    "                linea.append(str(value))  #appending the values in the list, after changing their type to strings, necessary for the columns formed by integers because if not done you can not concatenate the values.\n",
    "            report = report + \"\\n\" + \"\\t\".join(linea)   #adding a newline to reports to separete the previous row or line from the actual one. Then joined the values of the dicts with a tab as the separator.\n",
    "        docc = open(\"Report.tsv\", \"w\")     #opening the file in which I will pour the tables in write mode\n",
    "        docc.write(report)                 #writing in the file the report object comprised by the header and the values\n",
    "        docc.close()\n",
    "        \n",
    "        connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have used the INNER JOIN because both tables were joined in a 1:1 manner and there was no missing or null information in any of them. The way I have written the next data into the tsv file is simmilar to the one explained here therefore I won't go into as much detail in the next exercises regarding this aspect of the code."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a joined report that only includes the Genes SKOR and MAA3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus WHERE locus.Gene IN (\"SKOR\", \"MAA3\");\"\"\" #I am creating the joined table using sql syntax\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()\n",
    "        print(results)\n",
    "        report = \"\\n\" + \"\\t\".join(list(results[0].keys())) #creating the object that I will append to the end of the tsvfile\n",
    "        for dicc in results:\n",
    "            linea = []\n",
    "            for value in dicc.values():\n",
    "                linea.append(str(value))\n",
    "            report = report + \"\\n\" + \"\\t\".join(linea)\n",
    "        docc = open(\"Report.tsv\", \"a\")            #The opening mode of the document has been changed from write to append in order not to overwrite the file\n",
    "        docc.write(report)\n",
    "        docc.close()\n",
    "        \n",
    "        connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locus.Locus': 'AT1G01040', 'Gene': 'DCL1', 'ProteinLength': 332}, {'Locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locus.Locus': 'AT1G01060', 'Gene': 'LHY', 'ProteinLength': 290}, {'Locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locus.Locus': 'AT1G01140', 'Gene': 'CIPK9', 'ProteinLength': 223}, {'Locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locus.Locus': 'AT1G01220', 'Gene': 'FKGP', 'ProteinLength': 190}, {'Locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locus.Locus': 'AT2G03720', 'Gene': 'MRH6', 'ProteinLength': 189}, {'Locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locus.Locus': 'AT2G03800', 'Gene': 'GEK1', 'ProteinLength': 196}, {'Locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locus.Locus': 'AT2G04240', 'Gene': 'XERICO', 'ProteinLength': 256}, {'Locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locus.Locus': 'AT2G05210', 'Gene': 'POT1A', 'ProteinLength': 221}, {'Locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locus.Locus': 'AT3G02130', 'Gene': 'RPK2', 'ProteinLength': 284}, {'Locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locus.Locus': 'AT3G02140', 'Gene': 'TMAC2', 'ProteinLength': 300}, {'Locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locus.Locus': 'AT3G02230', 'Gene': 'RGP1', 'ProteinLength': 301}, {'Locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locus.Locus': 'AT3G02260', 'Gene': 'BIG', 'ProteinLength': 279}, {'Locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locus.Locus': 'AT3G02310', 'Gene': 'SEP2', 'ProteinLength': 175}, {'Locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locus.Locus': 'AT3G02680', 'Gene': 'NBS1', 'ProteinLength': 190}, {'Locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locus.Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProteinLength': 234}, {'Locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locus.Locus': 'AT3G02870', 'Gene': 'VTC4', 'ProteinLength': 311}, {'Locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locus.Locus': 'AT3G03260', 'Gene': 'HDG8', 'ProteinLength': 194}, {'Locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locus.Locus': 'AT4G14790', 'Gene': 'SUV3', 'ProteinLength': 312}, {'Locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locus.Locus': 'AT4G15210', 'Gene': 'BAM5', 'ProteinLength': 313}, {'Locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locus.Locus': 'AT4G15560', 'Gene': 'DXS', 'ProteinLength': 219}, {'Locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locus.Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProteinLength': 294}, {'Locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locus.Locus': 'AT4G15802', 'Gene': 'HSBP', 'ProteinLength': 254}, {'Locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locus.Locus': 'AT4G15880', 'Gene': 'ESD4', 'ProteinLength': 265}, {'Locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locus.Locus': 'AT4G16420', 'Gene': 'ADA2B', 'ProteinLength': 279}, {'Locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locus.Locus': 'AT4G16480', 'Gene': 'INT4', 'ProteinLength': 284}, {'Locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locus.Locus': 'AT5G10480', 'Gene': 'PAS2', 'ProteinLength': 301}, {'Locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locus.Locus': 'AT5G10510', 'Gene': 'AIL6', 'ProteinLength': 310}, {'Locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locus.Locus': 'AT5G11110', 'Gene': 'SPS2', 'ProteinLength': 232}, {'Locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locus.Locus': 'AT5G11260', 'Gene': 'HY5', 'ProteinLength': 221}, {'Locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locus.Locus': 'AT5G11510', 'Gene': 'MYB3R-4', 'ProteinLength': 336}, {'Locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locus.Locus': 'AT5G12200', 'Gene': 'PYD2', 'ProteinLength': 310}, {'Locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locus.Locus': 'AT5G13290', 'Gene': 'CRN', 'ProteinLength': 189}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus;\"\"\" #I am creating the joined table\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()\n",
    "        print(results)\n",
    "        codes = []                  #creating a list called codes\n",
    "        for dicc in results:        #iterating over the dictionaries in the dictionaries list results\n",
    "            codes.append(dicc[\"Locus\"])   #appending the values of the key Locus for all the dicts in the codes list \n",
    "        searchobj = \"\\t\".join(codes)      #joining the values of Locus code AIG with a tab value and saving it into the searcobj object which is a string\n",
    "        mo1 = re.findall( r'AT1G[0-9]+\\t', searchobj)    #doing a regular expresion search for the locus codes from each chromosome and saving it into a mo object. findall returns all the values that match the reg expr.\n",
    "        mo2 = re.findall( r'AT2G[0-9]+\\t', searchobj)\n",
    "        mo3 = re.findall( r'AT3G[0-9]+\\t', searchobj)\n",
    "        mo4 = re.findall( r'AT4G[0-9]+\\t', searchobj)\n",
    "        mo5 = re.findall( r'AT5G[0-9]+\\t?', searchobj)\n",
    "    \n",
    "    lista1 = [{\"Chromosome\" : \"1\", \"Chrom_entries\": len(mo1)}, {\"Chromosome\" : \"2\", \"Chrom_entries\": len(mo2)}, {\"Chromosome\" : \"3\", \"Chrom_entries\": len(mo3)}, {\"Chromosome\" : \"4\", \"Chrom_entries\": len(mo4)}, {\"Chromosome\" : \"5\", \"Chrom_entries\": len(mo5)}] #creating a list of dictionaries with the chromosome number and the number of entries that come from that chromosome.\n",
    "    report = \"\\n\" + \"\\t\".join(list(lista1[0].keys())) #same way of writting the values into the tsv file\n",
    "    for dicc in lista1:\n",
    "        linea = []\n",
    "        for value in dicc.values():\n",
    "            linea.append(str(value))\n",
    "        report = report + \"\\n\" + \"\\t\".join(linea)\n",
    "    docc = open(\"Report.tsv\", \"a\")\n",
    "    docc.write(report)\n",
    "    docc.close()\n",
    "        \n",
    "    connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locus.Locus': 'AT1G01040', 'Gene': 'DCL1', 'ProteinLength': 332}, {'Locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locus.Locus': 'AT1G01060', 'Gene': 'LHY', 'ProteinLength': 290}, {'Locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locus.Locus': 'AT1G01140', 'Gene': 'CIPK9', 'ProteinLength': 223}, {'Locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locus.Locus': 'AT1G01220', 'Gene': 'FKGP', 'ProteinLength': 190}, {'Locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locus.Locus': 'AT2G03720', 'Gene': 'MRH6', 'ProteinLength': 189}, {'Locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locus.Locus': 'AT2G03800', 'Gene': 'GEK1', 'ProteinLength': 196}, {'Locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locus.Locus': 'AT2G04240', 'Gene': 'XERICO', 'ProteinLength': 256}, {'Locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locus.Locus': 'AT2G05210', 'Gene': 'POT1A', 'ProteinLength': 221}, {'Locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locus.Locus': 'AT3G02130', 'Gene': 'RPK2', 'ProteinLength': 284}, {'Locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locus.Locus': 'AT3G02140', 'Gene': 'TMAC2', 'ProteinLength': 300}, {'Locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locus.Locus': 'AT3G02230', 'Gene': 'RGP1', 'ProteinLength': 301}, {'Locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locus.Locus': 'AT3G02260', 'Gene': 'BIG', 'ProteinLength': 279}, {'Locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locus.Locus': 'AT3G02310', 'Gene': 'SEP2', 'ProteinLength': 175}, {'Locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locus.Locus': 'AT3G02680', 'Gene': 'NBS1', 'ProteinLength': 190}, {'Locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locus.Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProteinLength': 234}, {'Locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locus.Locus': 'AT3G02870', 'Gene': 'VTC4', 'ProteinLength': 311}, {'Locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locus.Locus': 'AT3G03260', 'Gene': 'HDG8', 'ProteinLength': 194}, {'Locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locus.Locus': 'AT4G14790', 'Gene': 'SUV3', 'ProteinLength': 312}, {'Locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locus.Locus': 'AT4G15210', 'Gene': 'BAM5', 'ProteinLength': 313}, {'Locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locus.Locus': 'AT4G15560', 'Gene': 'DXS', 'ProteinLength': 219}, {'Locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locus.Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProteinLength': 294}, {'Locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locus.Locus': 'AT4G15802', 'Gene': 'HSBP', 'ProteinLength': 254}, {'Locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locus.Locus': 'AT4G15880', 'Gene': 'ESD4', 'ProteinLength': 265}, {'Locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locus.Locus': 'AT4G16420', 'Gene': 'ADA2B', 'ProteinLength': 279}, {'Locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locus.Locus': 'AT4G16480', 'Gene': 'INT4', 'ProteinLength': 284}, {'Locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locus.Locus': 'AT5G10480', 'Gene': 'PAS2', 'ProteinLength': 301}, {'Locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locus.Locus': 'AT5G10510', 'Gene': 'AIL6', 'ProteinLength': 310}, {'Locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locus.Locus': 'AT5G11110', 'Gene': 'SPS2', 'ProteinLength': 232}, {'Locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locus.Locus': 'AT5G11260', 'Gene': 'HY5', 'ProteinLength': 221}, {'Locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locus.Locus': 'AT5G11510', 'Gene': 'MYB3R-4', 'ProteinLength': 336}, {'Locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locus.Locus': 'AT5G12200', 'Gene': 'PYD2', 'ProteinLength': 310}, {'Locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locus.Locus': 'AT5G13290', 'Gene': 'CRN', 'ProteinLength': 189}]\n",
      "AT1G01040\tAT1G01060\tAT1G01140\tAT1G01220\tAT2G03720\tAT2G03800\tAT2G04240\tAT2G05210\tAT3G02130\tAT3G02140\tAT3G02230\tAT3G02260\tAT3G02310\tAT3G02680\tAT3G02850\tAT3G02870\tAT3G03260\tAT4G14790\tAT4G15210\tAT4G15560\tAT4G15570\tAT4G15802\tAT4G15880\tAT4G16420\tAT4G16480\tAT5G10480\tAT5G10510\tAT5G11110\tAT5G11260\tAT5G11510\tAT5G12200\tAT5G13290\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus;\"\"\" #I am creating the joined table\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()\n",
    "        print(results)\n",
    "        codes = []\n",
    "        for dicc in results:        #iterating over the dictionaries in the dictionaries list results\n",
    "            codes.append(dicc[\"Locus\"])   #appending \n",
    "        searchobj = \"\\t\".join(codes)\n",
    "        print(searchobj)\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                                user='root',\n",
    "                                password='root',\n",
    "                                db='exam_week_2',\n",
    "                                charset='utf8mb4', # note utf8... this is important for unusual characters!\n",
    "                                cursorclass=pymysql.cursors.DictCursor)\n",
    "#onnection.autocommit = False\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            sql = \"drop table chromosome\"\n",
    "            cursor.execute(sql)\n",
    "            loc = open(\"LocusGene.tsv\", \"r\")\n",
    "            locus = loc.readlines()\n",
    "            import re\n",
    "            mo = re.search( r'([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)[\\n]', locus[0])\n",
    "            colnames1 = [mo.group(1), mo.group(2), mo.group(3)]\n",
    "            sql = \"CREATE TABLE chromosome( \" +colnames1[0] + \" VARCHAR(20) NOT NULL PRIMARY KEY, \" +colnames1[1] + \" VARCHAR(30) NOT NULL, \" + colnames1[2] + \" INTEGER NOT NULL, chromnumber VARCHAR(2) NOT NULL  )\"\n",
    "            cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close() # if I close before I commit, the inserts are lostprint(colnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "7 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_exam_week_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('chromosome',), ('germplasm',), ('locus',)]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases\n",
    "%sql use exam_week_2\n",
    "%sql show tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            res = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            for row in res:\n",
    "                sql = \"INSERT INTO chromosome (Locus, Gene, ProteinLength, chromnumber ) values   ('{}', '{}',{}, '{}')\".format(row[colnames1[0]], row[colnames1[1]], row[colnames1[2]], row[colnames1[0]][2])\n",
    "                cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "        <th>chromnumber</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332, '1'),\n",
       " ('AT1G01060', 'LHY', 290, '1'),\n",
       " ('AT1G01140', 'CIPK9', 223, '1'),\n",
       " ('AT1G01220', 'FKGP', 190, '1'),\n",
       " ('AT2G03720', 'MRH6', 189, '2'),\n",
       " ('AT2G03800', 'GEK1', 196, '2'),\n",
       " ('AT2G04240', 'XERICO', 256, '2'),\n",
       " ('AT2G05210', 'POT1A', 221, '2'),\n",
       " ('AT3G02130', 'RPK2', 284, '3'),\n",
       " ('AT3G02140', 'TMAC2', 300, '3'),\n",
       " ('AT3G02230', 'RGP1', 301, '3'),\n",
       " ('AT3G02260', 'BIG', 279, '3'),\n",
       " ('AT3G02310', 'SEP2', 175, '3'),\n",
       " ('AT3G02680', 'NBS1', 190, '3'),\n",
       " ('AT3G02850', 'SKOR', 234, '3'),\n",
       " ('AT3G02870', 'VTC4', 311, '3'),\n",
       " ('AT3G03260', 'HDG8', 194, '3'),\n",
       " ('AT4G14790', 'SUV3', 312, '4'),\n",
       " ('AT4G15210', 'BAM5', 313, '4'),\n",
       " ('AT4G15560', 'DXS', 219, '4'),\n",
       " ('AT4G15570', 'MAA3', 294, '4'),\n",
       " ('AT4G15802', 'HSBP', 254, '4'),\n",
       " ('AT4G15880', 'ESD4', 265, '4'),\n",
       " ('AT4G16420', 'ADA2B', 279, '4'),\n",
       " ('AT4G16480', 'INT4', 284, '4'),\n",
       " ('AT5G10480', 'PAS2', 301, '5'),\n",
       " ('AT5G10510', 'AIL6', 310, '5'),\n",
       " ('AT5G11110', 'SPS2', 232, '5'),\n",
       " ('AT5G11260', 'HY5', 221, '5'),\n",
       " ('AT5G11510', 'MYB3R-4', 336, '5'),\n",
       " ('AT5G12200', 'PYD2', 310, '5'),\n",
       " ('AT5G13290', 'CRN', 189, '5')]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select * from chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>chromnumber</th>\n",
       "        <th>AVG(ProteinLength)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>258.7500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>215.5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>3</td>\n",
       "        <td>252.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "        <td>277.5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>5</td>\n",
       "        <td>271.2857</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('1', Decimal('258.7500')),\n",
       " ('2', Decimal('215.5000')),\n",
       " ('3', Decimal('252.0000')),\n",
       " ('4', Decimal('277.5000')),\n",
       " ('5', Decimal('271.2857'))]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT chromnumber, AVG(ProteinLength) FROM chromosome GROUP BY chromnumber"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'chromnumber': '1', 'AVG(ProteinLength)': Decimal('258.7500')}, {'chromnumber': '2', 'AVG(ProteinLength)': Decimal('215.5000')}, {'chromnumber': '3', 'AVG(ProteinLength)': Decimal('252.0000')}, {'chromnumber': '4', 'AVG(ProteinLength)': Decimal('277.5000')}, {'chromnumber': '5', 'AVG(ProteinLength)': Decimal('271.2857')}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "        \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT chromnumber, AVG(ProteinLength) FROM chromosome GROUP BY chromnumber;\"\"\" #I am creating the joined table\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()\n",
    "        print(results)\n",
    "        report = \"\\n\" + \"\\t\".join(list(results[0].keys())) #creating the header for the document\n",
    "        for dicc in results:\n",
    "            linea = []\n",
    "            for value in dicc.values():\n",
    "                linea.append(str(value))\n",
    "            report = report + \"\\n\" + \"\\t\".join(linea)\n",
    "        docc = open(\"Report.tsv\", \"a\")\n",
    "        docc.write(report)\n",
    "        docc.close()\n",
    "        \n",
    "        connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
