{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 1: controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re  #Importing the module for regular expressions\n",
    "germ = open(\"Germplasm.tsv\", \"r\") #Opening the Germplasm file in read mode and assigning it to an object\n",
    "germ.seek(0)  #Taking the cursor to the beginning of the file\n",
    "germlist = []\n",
    "for line in germ.readlines()[1:]:    #iterate over each line in the file which are string, except over the first line, index 0, that are the headers\n",
    "    print(line)\n",
    "    matchobj = re.search( r'^(AT[0-9]G[0-9]+)', line)#do a regular expression search for the locus codes and save the results in matchobj\n",
    "    germlist.append(matchobj.group(1)) #I create a list with all the locus codes to be able to compare it with the list from locusgene\n",
    "    print(\"The AGI locus codes are\", matchobj.group(1)) #print the results of the search for each line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re \n",
    "loc = open(\"LocusGene.tsv\", \"r\")\n",
    "loclist = []\n",
    "for line in loc.readlines()[1:]:\n",
    "    matchobj = re.search( r'^(AT[0-9]G[0-9]+)', line)\n",
    "    loclist.append(matchobj.group(1))\n",
    "    print(\"The AGI locus codes are\", matchobj.group(1))\n",
    "loc.seek(0)   # set it back to the beginning again for this lesson..\n",
    "loclist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if germlist == loclist:\n",
    "    print(\"Both documents follow the same locus sequence\")\n",
    "else:\n",
    "    print(\"The documents do not follow the same sequence\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I performed a regular expression search over the databases to check if both locus codes where written equally and then append them to a list to be able to check if they followed they same order. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loc = open(\"LocusGene.tsv\", \"r\")\n",
    "for line in loc.readlines():\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2: Design and create the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql create database exam_week_2;\n",
    "%sql show databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql use exam_week_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "germ = open(\"Germplasm.tsv\", \"r\")\n",
    "germplasm = germ.readlines()\n",
    "\n",
    "import re\n",
    "mo = re.search( r'([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)[\\n]', germplasm[0])\n",
    "colnames = [mo.group(1), mo.group(2), mo.group(3), mo.group(4)]\n",
    "colnames\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                                user='root',\n",
    "                                password='root',\n",
    "                                db='exam_week_2',\n",
    "                                charset='utf8mb4', # note utf8... this is important for unusual characters!\n",
    "                                cursorclass=pymysql.cursors.DictCursor)\n",
    "#onnection.autocommit = False\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv') as csvfile:\n",
    "            cursor.execute(\"drop table germplasm\")\n",
    "            germ = open(\"Germplasm.tsv\", \"r\")\n",
    "            germplasm = germ.readlines()\n",
    "            import re\n",
    "            mo = re.search( r'([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)[\\n]', germplasm[0])\n",
    "            colnames = [mo.group(1), mo.group(2), mo.group(3), mo.group(4)]\n",
    "            sql = \"CREATE TABLE germplasm( \" +colnames[0] + \" VARCHAR(12) NOT NULL PRIMARY KEY, \" +colnames[1] + \" VARCHAR(30) NOT NULL, \" + colnames[2] + \" VARCHAR(600) NOT NULL, \" + colnames[3] + \" INTEGER NOT NULL)\"\n",
    "            cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close() # if I close before I commit, the inserts are lostprint(colnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql drop table germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql select * from germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pymysql\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                                user='root',\n",
    "                                password='root',\n",
    "                                db='exam_week_2',\n",
    "                                charset='utf8mb4', # note utf8... this is important for unusual characters!\n",
    "                                cursorclass=pymysql.cursors.DictCursor)\n",
    "#onnection.autocommit = False\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            loc = open(\"LocusGene.tsv\", \"r\")\n",
    "            locus = loc.readlines()\n",
    "            import re\n",
    "            mo = re.search( r'([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)[\\n]', locus[0])\n",
    "            colnames1 = [mo.group(1), mo.group(2), mo.group(3)]\n",
    "            sql = \"CREATE TABLE locus( \" +colnames1[0] + \" VARCHAR(20) NOT NULL PRIMARY KEY, \" +colnames1[1] + \" VARCHAR(30) NOT NULL, \" + colnames1[2] + \" INTEGER NOT NULL)\"\n",
    "            cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close() # if I close before I commit, the inserts are lostprint(colnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%sql drop table locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql select * from locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv') as csvfile:\n",
    "            res = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            for row in res:\n",
    "                print(row)\n",
    "                sql = \"\"\"INSERT INTO germplasm (Locus, germplasm, phenotype, pubmed) values       \n",
    "                ('{}', '{}','{}',{})\"\"\".format(row[colnames[0]], row[colnames[1]], row[colnames[2]],row[colnames[3]]) #El de aqu√≠ es el que viene del csv y el de arriba el de las tablas sql\n",
    "                cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%sql select * from germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            res = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            for row in res:\n",
    "                print(row)\n",
    "                sql = \"\"\"INSERT INTO locus (Locus, Gene, ProteinLength) values       \n",
    "                ('{}', '{}',{})\"\"\".format(row[colnames1[0]], row[colnames1[1]], row[colnames1[2]]) #El de aqu√≠ es el que viene del csv y el de arriba el de las tablas sql\n",
    "                cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql select * from locus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "import io\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus;\"\"\" #I am creating the joined table\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()            #saving the table in the object results which is a list of dictionaries\n",
    "        report = \"\\t\".join(list(results[0].keys())) #creating the header for the document\n",
    "        for dicc in results:\n",
    "            linea = []\n",
    "            for value in dicc.values():\n",
    "                linea.append(str(value))\n",
    "            report = report + \"\\n\" + \"\\t\".join(linea)\n",
    "        docc = open(\"Report.tsv\", \"w\")\n",
    "        docc.write(report)\n",
    "        docc.close()\n",
    "        \n",
    "        connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a joined report that only includes the Genes SKOR and MAA3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus WHERE locus.Gene IN (\"SKOR\", \"MAA3\");\"\"\" #I am creating the joined table\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()\n",
    "        print(results)\n",
    "        report = \"\\n\" + \"\\t\".join(list(results[0].keys())) #creating the header for the document\n",
    "        for dicc in results:\n",
    "            linea = []\n",
    "            for value in dicc.values():\n",
    "                linea.append(str(value))\n",
    "            report = report + \"\\n\" + \"\\t\".join(linea)\n",
    "        docc = open(\"Report.tsv\", \"a\")\n",
    "        docc.write(report)\n",
    "        docc.close()\n",
    "        \n",
    "        connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locus.Locus': 'AT1G01040', 'Gene': 'DCL1', 'ProteinLength': 332}, {'Locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locus.Locus': 'AT1G01060', 'Gene': 'LHY', 'ProteinLength': 290}, {'Locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locus.Locus': 'AT1G01140', 'Gene': 'CIPK9', 'ProteinLength': 223}, {'Locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locus.Locus': 'AT1G01220', 'Gene': 'FKGP', 'ProteinLength': 190}, {'Locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locus.Locus': 'AT2G03720', 'Gene': 'MRH6', 'ProteinLength': 189}, {'Locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locus.Locus': 'AT2G03800', 'Gene': 'GEK1', 'ProteinLength': 196}, {'Locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locus.Locus': 'AT2G04240', 'Gene': 'XERICO', 'ProteinLength': 256}, {'Locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locus.Locus': 'AT2G05210', 'Gene': 'POT1A', 'ProteinLength': 221}, {'Locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locus.Locus': 'AT3G02130', 'Gene': 'RPK2', 'ProteinLength': 284}, {'Locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locus.Locus': 'AT3G02140', 'Gene': 'TMAC2', 'ProteinLength': 300}, {'Locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locus.Locus': 'AT3G02230', 'Gene': 'RGP1', 'ProteinLength': 301}, {'Locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locus.Locus': 'AT3G02260', 'Gene': 'BIG', 'ProteinLength': 279}, {'Locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locus.Locus': 'AT3G02310', 'Gene': 'SEP2', 'ProteinLength': 175}, {'Locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locus.Locus': 'AT3G02680', 'Gene': 'NBS1', 'ProteinLength': 190}, {'Locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locus.Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProteinLength': 234}, {'Locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locus.Locus': 'AT3G02870', 'Gene': 'VTC4', 'ProteinLength': 311}, {'Locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locus.Locus': 'AT3G03260', 'Gene': 'HDG8', 'ProteinLength': 194}, {'Locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locus.Locus': 'AT4G14790', 'Gene': 'SUV3', 'ProteinLength': 312}, {'Locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locus.Locus': 'AT4G15210', 'Gene': 'BAM5', 'ProteinLength': 313}, {'Locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locus.Locus': 'AT4G15560', 'Gene': 'DXS', 'ProteinLength': 219}, {'Locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locus.Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProteinLength': 294}, {'Locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locus.Locus': 'AT4G15802', 'Gene': 'HSBP', 'ProteinLength': 254}, {'Locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locus.Locus': 'AT4G15880', 'Gene': 'ESD4', 'ProteinLength': 265}, {'Locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locus.Locus': 'AT4G16420', 'Gene': 'ADA2B', 'ProteinLength': 279}, {'Locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locus.Locus': 'AT4G16480', 'Gene': 'INT4', 'ProteinLength': 284}, {'Locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locus.Locus': 'AT5G10480', 'Gene': 'PAS2', 'ProteinLength': 301}, {'Locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locus.Locus': 'AT5G10510', 'Gene': 'AIL6', 'ProteinLength': 310}, {'Locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locus.Locus': 'AT5G11110', 'Gene': 'SPS2', 'ProteinLength': 232}, {'Locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locus.Locus': 'AT5G11260', 'Gene': 'HY5', 'ProteinLength': 221}, {'Locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locus.Locus': 'AT5G11510', 'Gene': 'MYB3R-4', 'ProteinLength': 336}, {'Locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locus.Locus': 'AT5G12200', 'Gene': 'PYD2', 'ProteinLength': 310}, {'Locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locus.Locus': 'AT5G13290', 'Gene': 'CRN', 'ProteinLength': 189}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasm INNER JOIN locus ON germplasm.Locus = locus.Locus;\"\"\" #I am creating the joined table\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()\n",
    "        print(results)\n",
    "        codes = []\n",
    "        for dicc in results:\n",
    "            codes.append(dicc[\"Locus\"])\n",
    "        searchobj = \"\\t\".join(codes)\n",
    "        mo1 = re.findall( r'AT1G[0-9]+\\t', searchobj)\n",
    "        mo2 = re.findall( r'AT2G[0-9]+\\t', searchobj)\n",
    "        mo3 = re.findall( r'AT3G[0-9]+\\t', searchobj)\n",
    "        mo4 = re.findall( r'AT4G[0-9]+\\t', searchobj)\n",
    "        mo5 = re.findall( r'AT5G[0-9]+\\t?', searchobj)\n",
    "    \n",
    "    lista1 = [{\"Chromosome\" : \"1\", \"Chrom_entries\": len(mo1)}, {\"Chromosome\" : \"2\", \"Chrom_entries\": len(mo2)}, {\"Chromosome\" : \"3\", \"Chrom_entries\": len(mo3)}, {\"Chromosome\" : \"4\", \"Chrom_entries\": len(mo4)}, {\"Chromosome\" : \"5\", \"Chrom_entries\": len(mo5)}]\n",
    "    report = \"\\n\" + \"\\t\".join(list(lista1[0].keys())) #creating the header for the document\n",
    "    for dicc in lista1:\n",
    "        linea = []\n",
    "        for value in dicc.values():\n",
    "            linea.append(str(value))\n",
    "        report = report + \"\\n\" + \"\\t\".join(linea)\n",
    "    docc = open(\"Report.tsv\", \"a\")\n",
    "    docc.write(report)\n",
    "    docc.close()\n",
    "        \n",
    "    connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                                user='root',\n",
    "                                password='root',\n",
    "                                db='exam_week_2',\n",
    "                                charset='utf8mb4', # note utf8... this is important for unusual characters!\n",
    "                                cursorclass=pymysql.cursors.DictCursor)\n",
    "#onnection.autocommit = False\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            sql = \"drop table chromosome\"\n",
    "            cursor.execute(sql)\n",
    "            loc = open(\"LocusGene.tsv\", \"r\")\n",
    "            locus = loc.readlines()\n",
    "            import re\n",
    "            mo = re.search( r'([^\\t]+)\\t([^\\t]+)\\t([^\\t]+)[\\n]', locus[0])\n",
    "            colnames1 = [mo.group(1), mo.group(2), mo.group(3)]\n",
    "            sql = \"CREATE TABLE chromosome( \" +colnames1[0] + \" VARCHAR(20) NOT NULL PRIMARY KEY, \" +colnames1[1] + \" VARCHAR(30) NOT NULL, \" + colnames1[2] + \" INTEGER NOT NULL, chromnumber VARCHAR(2) NOT NULL  )\"\n",
    "            cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close() # if I close before I commit, the inserts are lostprint(colnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "7 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_exam_week_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('chromosome',), ('germplasm',), ('locus',)]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases\n",
    "%sql use exam_week_2\n",
    "%sql show tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "    \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            res = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            for row in res:\n",
    "                sql = \"INSERT INTO chromosome (Locus, Gene, ProteinLength, chromnumber ) values   ('{}', '{}',{}, '{}')\".format(row[colnames1[0]], row[colnames1[1]], row[colnames1[2]], row[colnames1[0]][2])\n",
    "                cursor.execute(sql)\n",
    "            connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "        <th>chromnumber</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "        <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "        <td>5</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332, '1'),\n",
       " ('AT1G01060', 'LHY', 290, '1'),\n",
       " ('AT1G01140', 'CIPK9', 223, '1'),\n",
       " ('AT1G01220', 'FKGP', 190, '1'),\n",
       " ('AT2G03720', 'MRH6', 189, '2'),\n",
       " ('AT2G03800', 'GEK1', 196, '2'),\n",
       " ('AT2G04240', 'XERICO', 256, '2'),\n",
       " ('AT2G05210', 'POT1A', 221, '2'),\n",
       " ('AT3G02130', 'RPK2', 284, '3'),\n",
       " ('AT3G02140', 'TMAC2', 300, '3'),\n",
       " ('AT3G02230', 'RGP1', 301, '3'),\n",
       " ('AT3G02260', 'BIG', 279, '3'),\n",
       " ('AT3G02310', 'SEP2', 175, '3'),\n",
       " ('AT3G02680', 'NBS1', 190, '3'),\n",
       " ('AT3G02850', 'SKOR', 234, '3'),\n",
       " ('AT3G02870', 'VTC4', 311, '3'),\n",
       " ('AT3G03260', 'HDG8', 194, '3'),\n",
       " ('AT4G14790', 'SUV3', 312, '4'),\n",
       " ('AT4G15210', 'BAM5', 313, '4'),\n",
       " ('AT4G15560', 'DXS', 219, '4'),\n",
       " ('AT4G15570', 'MAA3', 294, '4'),\n",
       " ('AT4G15802', 'HSBP', 254, '4'),\n",
       " ('AT4G15880', 'ESD4', 265, '4'),\n",
       " ('AT4G16420', 'ADA2B', 279, '4'),\n",
       " ('AT4G16480', 'INT4', 284, '4'),\n",
       " ('AT5G10480', 'PAS2', 301, '5'),\n",
       " ('AT5G10510', 'AIL6', 310, '5'),\n",
       " ('AT5G11110', 'SPS2', 232, '5'),\n",
       " ('AT5G11260', 'HY5', 221, '5'),\n",
       " ('AT5G11510', 'MYB3R-4', 336, '5'),\n",
       " ('AT5G12200', 'PYD2', 310, '5'),\n",
       " ('AT5G13290', 'CRN', 189, '5')]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select * from chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>chromnumber</th>\n",
       "        <th>AVG(ProteinLength)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>258.7500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>215.5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>3</td>\n",
       "        <td>252.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "        <td>277.5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>5</td>\n",
       "        <td>271.2857</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('1', Decimal('258.7500')),\n",
       " ('2', Decimal('215.5000')),\n",
       " ('3', Decimal('252.0000')),\n",
       " ('4', Decimal('277.5000')),\n",
       " ('5', Decimal('271.2857'))]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT chromnumber, AVG(ProteinLength) FROM chromosome GROUP BY chromnumber"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'chromnumber': '1', 'AVG(ProteinLength)': Decimal('258.7500')}, {'chromnumber': '2', 'AVG(ProteinLength)': Decimal('215.5000')}, {'chromnumber': '3', 'AVG(ProteinLength)': Decimal('252.0000')}, {'chromnumber': '4', 'AVG(ProteinLength)': Decimal('277.5000')}, {'chromnumber': '5', 'AVG(ProteinLength)': Decimal('271.2857')}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_week_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "        \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT chromnumber, AVG(ProteinLength) FROM chromosome GROUP BY chromnumber;\"\"\" #I am creating the joined table\n",
    "        cursor.execute(sql) \n",
    "        results = cursor.fetchall()\n",
    "        print(results)\n",
    "        report = \"\\n\" + \"\\t\".join(list(results[0].keys())) #creating the header for the document\n",
    "        for dicc in results:\n",
    "            linea = []\n",
    "            for value in dicc.values():\n",
    "                linea.append(str(value))\n",
    "            report = report + \"\\n\" + \"\\t\".join(linea)\n",
    "        docc = open(\"Report.tsv\", \"a\")\n",
    "        docc.write(report)\n",
    "        docc.close()\n",
    "        \n",
    "        connection.commit() # NOTE THAT I AM FORCING THE WRITE TO THE DATABASE HERE\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
